File: LeetCode-DSA/LeetCode75/Level1/BinarySearch.java
/*
Q704. Description: Given an array of integers nums which is sorted in ascending order, 
and an integer target, write a function to search target in nums. 
If target exists, then return its index. Otherwise, return -1.
You must write an algorithm with O(log n) runtime complexity.
*/

class Solution {
    public int search(int[] nums, int target) 
    {
        int start = 0;
        int end = nums.length - 1;
        int mid = 0;

        while(start <= end)
        {
            // If element found at middle, return index
            mid = (start + end)/2;
            if(target == nums[mid])
                return mid;
            // If number is greater than middle, search in upper half
            else if(target > nums[mid])
                start = mid + 1;
            else
                end = mid - 1;
        }
        // Element not found anywhere in list
        return -1;
    }
}

File: LeetCode-DSA/Algorithms1/BinarySearch.cpp
// Q704. Binary search in ascending order array
class Solution {
public:
    int search(vector<int>& nums, int target) {
        // Initialize
        int start = 0;
        int end = nums.size() - 1;
        int mid = 0;
        while(start <= end)
        {
            mid = start + (end - start) / 2;
            if(nums[mid] == target)
                return mid;
            else if(nums[mid] < target)
            {
                // Search in right side from mid+1 to end of array
                start = mid + 1;
            }
            else
            {
                // Search in left side from 0 to mid-1
                end = mid - 1;
            }
        }
        // Element not found
        return -1;
    }
};